### ISSUE TITLE
Tracking packages installed from sources and assigning CVE to them (Packages not present in RPM database)

### ISSUE THREAD

**Issue by mradaszewski92** [wazuh/wazuh#11316]
|Wazuh version|Component|
|---|---|
| 4.2.1 | Wazuh Vulnerability-detector|

Hello Wazuh Team,

is there a way to gather information, track and checksums of applications that are installed from source and their packages are not visible in the RPM database? The idea is also to capture the checksums of vulnerable packages and assign the correct CVE to them

**Comment by chemamartinez** [comment]
Hi @mradaszewski92,

This is an interesting use case to discuss how to extend the current scope of vulnerability detection in Wazuh.

I have a question about your purpose. It would help me to better understand your goal if you could answer how to correlate CVEs with package checksums. Currently, the Vulnerability Detector is mainly based on comparing versions between installed packages and official feeds such as the NVD.

[Here](https://documentation.wazuh.com/current/user-manual/capabilities/vulnerability-detection/how-it-works.html#how-it-works) you can find the source of every feed collected by Wazuh to detect vulnerabilities. None of these feeds is based on the checksum to match vulnerabilities.

On the other hand, unfortunately, the agent is not capable of inventory packages that are not visible in the RPM or DEB databases in RPM or DEB Linux-based OS. This capability could be implemented in future versions.

As you mentioned, FIM is the appropriate component to create an inventory of checksums for any particular set of files. Even alerts are not generated until changes happen to these files, the first scan generates a baseline and stores in the agent database all the metadata for the monitored files.

You can find that inventory in the Wazuh UI as you can see in this screenshot.
![image](https://user-images.githubusercontent.com/29475387/145837251-a6ad7829-481e-43cb-9f35-f956960e404c.png)
They are also stored in SQLite databases on the manager side, in the location `/var/ossec/queue/db/agentID.db`.

I hope this information helps! Don't hesitate to ask for further questions.

**Comment by mradaszewski92** [comment]
correct, but currently there is no such information in the elasticsearch index. Considering the last 0day, it was extremely important to track log4j packages in a short time

I checked and actually after restarting the agent, I can see the files being searched for. And now my point is that it could be found from the elasticsearch level

**Comment by xufengnian** [comment]
I have encountered a similar problem with how to detect log4j2 related vulnerabilities or find log4j2 packages with wazuh.
The derivation is how wazuh can support the detection of dependency files. As a good HIDS project, I think wazuh would be stronger if it could support the above features.

I have referred to other HIDS products, some of them include "asset inventory" function, which can inventory jar packages, so that when there is a vulnerability, you can quickly find the problematic host by searching
Take jar package inventory as an example, some HIDS periodically iterate through all the current java processes, find the directory where the process is located, and then use the regular way to match all the jar packages in the directory, so that you can find the relevant dependency files being used in a timely manner, the current wazuh inventory with FIM should be able to achieve this function, thank you very much!

**Comment by chemamartinez** [comment]
Hi @mradaszewski92 and @xufengnian,

Thank you both for your feedback!

As I told you before, currently is hard for Vulnerability Detector to detect this CVE since it affects a library instead of a simple application, so the vulnerability sources such as the NVD are not precise with the affected packages.
See https://nvd.nist.gov/vuln/detail/CVE-2021-44228.

On the other hand, we are continuously improving the inventory and vulnerability detection capabilities so I hope in the near future to be able to gather JAR packages and extend the detection scope.

**Comment by mradaszewski92** [comment]
is there any chance to get some feedback when the above ideas are found on your road map?

**Comment by xufengnian** [comment]
> Hi @mradaszewski92 and @xufengnian,
> 
> Thank you both for your feedback!
> 
> As I told you before, currently is hard for Vulnerability Detector to detect this CVE since it affects a library instead of a simple application, so the vulnerability sources such as the NVD are not precise with the affected packages. See https://nvd.nist.gov/vuln/detail/CVE-2021-44228.
> 
> On the other hand, we are continuously improving the inventory and vulnerability detection capabilities so I hope in the near future to be able to gather JAR packages and extend the detection scope.

Chinese network security vendor qingteng has a commercial HIDS, many years ago with an inventory of jar packages, web sites, databases, hardware configuration, port processes and other functions, I think wazuh can also refer to this piece of product screening jar packages is very convenient, the following picture can be illustrated, do not need to read Chinese, the meaning of this picture is very simple, for qingteng HIDS, you only need to enter the keyword "log4j" on the server side, you can search the relevant jar package host, version, path information, 
![image](https://user-images.githubusercontent.com/15324125/145991937-a9a769b2-6acb-4fd3-b372-a5ca99a60ae5.png)

their official website also has an English introduction, I think it is very good, see 
https://www.qingteng.cn/en/asset-inventory.html

**Comment by chemamartinez** [comment]
Thanks again for your feedback! @xufengnian we will definitely take a look into that HIDS.

We already have a similar feature request in our roadmap (https://github.com/wazuh/wazuh/issues/8750), where it was requested to collect packages apart from the ones installed via deb or rpm.

We hope we can start to work on this feature ASAP. The progress can be tracked by the issues of this repository.

Best regards,
Chema.

**Comment by xufengnian** [comment]
Thank you very much, I have followed this new feature and look forward to wazuh getting better and better!

**Comment by xufengnian** [comment]
This is another jar package scan python script for log4j2,I think maybe is useful for wazuh
https://github.com/lijiejie/log4j2_vul_local_scanner

**Comment by mradaszewski92** [comment]
hi, what is the chance that the functionality we are applying for will be in RC for Wazuh 4.3.0 / 4.4.0? :)
